barretenberg/cpp/src/barretenberg/ultra_honk/databus.test.cpp

```diff
@@ -0,0 +1,166 @@
+#include <cstddef>
+#include <cstdint>
+#include <gtest/gtest.h>
+
+#include "barretenberg/common/log.hpp"
+#include "barretenberg/goblin/mock_circuits.hpp"
+#include "barretenberg/plonk_honk_shared/instance_inspector.hpp"
+#include "barretenberg/stdlib_circuit_builders/goblin_ultra_circuit_builder.hpp"
+#include "barretenberg/stdlib_circuit_builders/ultra_circuit_builder.hpp"
+
+#include "barretenberg/ultra_honk/ultra_prover.hpp"
+#include "barretenberg/ultra_honk/ultra_verifier.hpp"
+
+using namespace bb;
+
+namespace {
+auto& engine = numeric::get_debug_randomness();
+}
+
+class DataBusTests : public ::testing::Test {
```

Comment:
This is a great suggestion and I did think about this but I struggled (and am still struggling) to come up with a way to do it. As you may have noticed, I was very deliberate in limiting the interface to the databus via access specifiers so I'm actually not sure it's possible to create a failing circuit. The only interface is read_bus_vector and if you try to read an invalid index, it will trigger an ASSERT. If you read a valid index, the read gate will be valid by definition. I'll keep thinking about this but if its really the case that an invalid gate can't be created then that seems like a good thing.

Reply:
My position is that if our interfaces restrict us writing failure tests then our interfaces are wrong. We either need to use friendship (of the issue is access specifiers) or we need to change the interface.